fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadReportingServiceImpl.java

```diff
@@ -161,16 +167,48 @@ public GenericResultsetData retrieveGenericResultset(final String name, final St
 
         final long startTime = System.currentTimeMillis();
         LOG.info("STARTING REPORT: {}   Type: {}", name, type);
-
-        final String sql = getSQLtoRun(name, type, queryParams, isSelfServiceUserReport);
-
-        final GenericResultsetData result = this.genericDataService.fillGenericResultSet(sql);
+        final StringBuilder sqlStringBuilder = new StringBuilder(200);
+        sqlStringBuilder.append(getSQLtoRun(name, type, queryParams, isSelfServiceUserReport));
+        final GenericResultsetData result;
+        boolean isPaginationAllowed = Boolean.parseBoolean(queryParams.get(ReportingConstants.IS_PAGINATION_ALLOWED));
+
+        if (isPaginationAllowed) {
+            result = retrieveGenericResultsetWithPagination(sqlStringBuilder, queryParams);
+        } else {
+            result = this.genericDataService.fillGenericResultSet(sqlStringBuilder.toString());
+        }
 
         final long elapsed = System.currentTimeMillis() - startTime;
         LOG.info("FINISHING Report/Request Name: {} - {}     Elapsed Time: {}", name, type, elapsed);
         return result;
     }
 
+    public GenericResultsetData retrieveGenericResultsetWithPagination(final StringBuilder sqlStringBuilder,
+            final Map<String, String> queryParams) {
+        final GenericResultsetData result;
+        final List<ApiParameterError> dataValidationErrors = new ArrayList<>();
+
+        final DataValidatorBuilder baseDataValidator = new DataValidatorBuilder(dataValidationErrors);
```

Comment:
Please extract the validation to a separate method/class to improve readability.